### ISSUE TITLE
Inconsistent calculations in CategoricalCrossentropy using logits

### ISSUE THREAD

**Issue by Frightera** [keras-team/keras#17634]
**Describe the problem**.

It seems like a bug in the calculation of cross entropy while calculating the log.

**Describe the current behavior**.
If the logits get too big, the calculation of `CategoricalCrossEntropy` is effected. We expect two of them to be same:

```python
logits = tf.constant([[14.4, 10.1, 3.5]])
y_true = tf.constant([[1, 0, 0]], dtype=tf.float32)

cce_obj = keras.losses.CategoricalCrossentropy(
                      from_logits=True)(y_true, logits)

cce_obj_logits = keras.losses.CategoricalCrossentropy(
                      from_logits=False)(y_true, tf.nn.softmax(logits))
```

which returns the same value, `0.01349`.

If the logits get bigger, they will return different values.

**Describe the expected behavior**.
Those two losses should produce the same result.

- Do you want to contribute a PR? (yes/no): Yes. Thim seems like an issue while calculating the log in categorical cross entropy losses.

**Standalone code to reproduce the issue**.
Please check the [gist here](https://colab.research.google.com/gist/Frightera/3959f54c0d06d69b0683161032c29c4f/cce-inconsistent-results.ipynb
). 

Edit (13/03/2023): 

Here are the results when logits are _larger_:

```python
logits = tf.constant([[14.4, 50.1, 3.5]], dtype = tf.float64)
y_true = tf.constant([[1, 0, 0]], dtype=tf.float64)

cce_obj = keras.losses.CategoricalCrossentropy(
                      from_logits=True)

loss = cce_obj(y_true, logits)
loss.numpy() # 35.7

cce_obj = keras.losses.CategoricalCrossentropy(
                      from_logits=False)

loss = cce_obj(y_true, tf.nn.softmax(logits))
loss.numpy() # 16.11
```

**Comment by tilakrayal** [comment]
@Frightera,
This may be expected because the cross entropy operation is unstable and a suggestion would be to try with **float64** everywhere. Thank you!

**Comment by Frightera** [comment]
Hi @tilakrayal,

For convention I'll copy paste what I mean:

```
import tensorflow as tf
from tensorflow import keras

logits = tf.constant([[14.4, 50.1, 3.5]], dtype=tf.float64)
y_true = tf.constant([[1, 0, 0]], dtype=tf.float64)

cce_obj = keras.losses.CategoricalCrossentropy(
                      from_logits=True)

loss = cce_obj(y_true, logits)
loss.numpy() # Returns 35.7

## -- Testing Keras backend calculation

cce_obj = keras.losses.CategoricalCrossentropy(
                      from_logits=False)

loss = cce_obj(y_true, tf.nn.softmax(logits))
loss.numpy() # Returns 16.11
```

So it has nothing to do with `tf.float64`, please check the gist for the comparison, thanks.

**Comment by Frightera** [comment]
Tracking down the issue, it is because of the values are clipped:

https://github.com/keras-team/keras/blob/0468c5cfd09a4d4bed7b810b58c14c3278231120/keras/backend.py#L5572-L5574

**Comment by Frightera** [comment]
@tilakrayal 

I think you missed the above comments. Could you please add `team-review-pending` tag so this can be further discussed?

**Comment by tilakrayal** [comment]
@sachinprasadhs,
I was able to reproduce the issue on tensorflow v2.11 and tf-nightly. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/da7ebcf8d8abfce785859d5b7961410d/untitled1037.ipynb).

**Comment by Frightera** [comment]
@sachinprasadhs 

Is there any specific reason why `1e-7` is used as `epsilon` for cross entropy losses? 

This creates confusion from time to time: https://stackoverflow.com/questions/76494166/why-tf-keras-losses-categoricalcrossentropy-result-different-to-tf-nn-softmax-cr/76495702

Maybe this value can be changed for CE losses.

**Comment by sachinprasadhs** [comment]
`1e-7` epsilon value is added as it is in the acceptable range of precession errors and it does not cause big difference in numerical computations. 
Refer the comment here for more details https://github.com/tensorflow/tensorflow/issues/60113#issuecomment-1504081934
